pkg/apis/batch/types.go

```diff
@@ -340,6 +356,19 @@ type JobSpec struct {
 	//
 	// +optional
 	Suspend *bool
+
+	// podReplacementPolicy specifies when to create replacement Pods. Possible values are:
+	// - TerminatingOrFailed means to create a replacement Pod when the previously
+	//   created Pod is terminating or failed.
+	// - Failed means to wait until a previously created Pod is fully terminated
+	//   before creating a replacement Pod.
+	//
+	// When using podFailurePolicy, the default value is Failed and this is the
+	// only allowed policy.
+	// When not using podFailurePolicy, the default value is TerminatingOrFailed.
+	// This is an alpha feature so you can enable JobPodReplacementPolicy for testing.
+	// +optional
+	PodReplacementPolicy *PodReplacementPolicy
```

Comment:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3939-allow-replacement-when-fully-terminated#pods-are-not-guaranteed-to-transition-to-a-terminal-phase

Maybe @mimowo can answer this also?

Reply:
And https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3939-allow-replacement-when-fully-terminated#pods-are-not-guaranteed-to-transition-to-a-terminal-phase has also another

Reply:
I don't think we have a default, as discussed earlier, so I think this section can be removed or rephrased. What we have is that when podFailurePolicy is specified the pods are replaced after reaching terminal phase - as if Failed was used.

Reply:
I think its a default though..  If PodFailurePolicy is specified we only allow the value of Failed but if you don't specify it we will interpret nil as Failed.  

That to me means that we technically have a default.  We just don't store in the spec as a default value due to api requirements for new fields.

Reply:
@deads2k pointed out that we should add a default.  Added defaulting this so it makes it a bit clear.

Reply:
I agree that it is useful to have `Failed` by default when using `podFailurePolicy`, but can we be sure nobody will want a `podFailurePolicy` with `TerminatingOrFailed`? Eg with `OnPodConditions` or with a future enhancement to the `podFailurePolicy`? I suppose we can relax validation for that in the future if needed.

We should properly document this, if everyone is okay with the confusing defaulting (from the API perspective)

Reply:
This decision was an outcome of the KEP review. The conclusion was that, when using podFailurePolicy, it doesn't make sense to have replacementPolicy=TerminatingOrFailed.

If you replace a terminating pod, you are ignoring the podFailurePolicy. PodFailurePolicies can only be evaluated for fully terminated Pods (otherwise you don't have the necessary information in the status).
As a result, you could be violating the policy by creating a replacement pod.